src/Dotnet.Script/Program.cs

```diff
@@ -232,20 +238,104 @@ private static int Wain(string[] args)
+                        {
+                            optimizationLevel = OptimizationLevel.Release;
+                        }
+                        exitCode = await RunScript(file.Value, debugMode.HasValue(), logFactory, optimizationLevel, app.RemainingArguments.Concat(argsAfterDoubleHypen), interactive.HasValue(), packageSources.Values?.ToArray());
+                    }
+                    else
+                    {
+                        string cacheFolder = Path.Combine(Path.GetTempPath(), "dotnet-scripts");
+                        // create unique folder name based on the path
+                        string uniqueFolderName = "";
+                        using (var sha = SHA256.Create())
+                        {
+                            uniqueFolderName = Convert.ToBase64String(sha.ComputeHash(Encoding.Unicode.GetBytes(file.Value))).Replace("=", String.Empty).Replace("/", string.Empty);
+                        }
+
+                        string publishDirectory = Path.Combine(cacheFolder, uniqueFolderName);
+                        if (!Directory.Exists(publishDirectory))
+                        {
+                            Directory.CreateDirectory(publishDirectory);
+                        }
+
+                        string absoluteSourcePath;
+                        SourceText code;
+                        if (!File.Exists(file.Value))
+                        {
+                            if (IsHttpUri(file.Value))
+                            {
+                                var downloader = new ScriptDownloader();
+                                var rawCode = await downloader.Download(file.Value);
+                                absoluteSourcePath = Path.Combine(publishDirectory, "source.csx");
+                                File.WriteAllText(absoluteSourcePath, rawCode);
+                                code = SourceText.From(rawCode);
+                            }
+                            else
+                            {
+                                throw new Exception($"Couldn't find file '{file}'");
+                            }
+                        }
+                        else
+                        {
+                            absoluteSourcePath = Path.IsPathRooted(file.Value) ? Path.GetFullPath(file.Value) : Path.GetFullPath(Path.Combine(Directory.GetCurrentDirectory(), file.Value));
+                            code = SourceText.From(File.ReadAllText(absoluteSourcePath));
+                        }
+
+                        // given the path to a script we create a %temp%\dotnet-scripts\{uniqueFolderName} path
+                        string pathToDll = Path.Combine(publishDirectory, Path.GetFileNameWithoutExtension(absoluteSourcePath) + ".dll");
+
+                        // source hash is the checkSum of the code
+                        string sourceHash = Convert.ToBase64String(code.GetChecksum().ToArray());
```

Comment:
The only thing that I am worried about here is NuGet packages referenced without version numbers. 
```
#r "nuget: SomePackage"
```
At some point `SomePackage` gets bumped while we still run with the old version.

Reply:
I think this is an acceptable tradeoff. 
I also think at some point in the future we will introduce some global JSON settings file so you could preset certain flags (i.e. --no-cache) globally so that should help

Reply:
Should we at the very least offer a way to clear the cache? 

`dotnet script clear-cache` 

I don't know if that should clear all cached scripts or just the "current" one ?

Reply:
@filipw thoughts?

Reply:
it also seems like something that more clever caching of the project file would allow us to decide that it is time to recompile.

Reply:
I have updated the README to describe the dll caching behavior